### PR TITLE
WIP: Sketch of comm event performance test

### PR DESCRIPTION
**Pull request by JoshOrndorff**
## Overview
This rholang contract is intended to create lots of comm events to test the performance of RSpace. A truly elegant solution is not possible due to limitations in the expressivity of the current rholang language (see comments at beginning of file, and discord post in pub-RChain#rholang from 7 Mar 2019). Nevertheless, this is a start.

### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-2839

### Notes
I would really love to add explicit substitution to rholang.

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [ ] includes tests for all added features,
- [ ] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by JoshOrndorff** [comment]
Here's the version that I would really love to use which almost entirely eliminates bottlenecking on channels. The explicit substitution is what's missing IMHO.

```scala
new adam,
    adamWomb,
    stdout(`rho:io:stdout`) in {

  /**
   * Template for a new person.
   * Persistent send in _not_ needed in this case because each offspring will have its
   * own copy. They will not need to contend for this original channel.
   * The process sent on adamWomb has free variables individual and individulaWomb.
   * The process sent on adamWomb is identical to the adam code below, and is
   * commented down there, not here.
   */
  adamWomb!(
    for (@generations <- individual) {
      for (newPerson <- individualWomb) {
        new cane,
            abel,
            caneWomb,
            abelWomb in {

          caneWomb!(*newPersonTemplate) |
          abelWomb!(*newPersonTemplate) |

          cane!(generations - 1) |
          abel!(generations - 1) |

          *newPerson[cane / individual][caneWomb / individualWomb] |
          *newPerson[abel / individual][abelWomb / individualWomb]
        }
      }
    }
  ) |


  // Same code as the template for the new person,
  // except adam's channels hardcoded instead of free.
  for (@generations <- adam) {
    for (newPerson <- adamWomb) {
      new cane,
          abel,
          caneWomb,
          abelWomb in {

        // Make sure all offspring can, themselves, reproduce.
        // Because each individual has its own womb, there is no
        // only a bottleneck among direct siblings.
        caneWomb!(*newPersonTemplate) |
        abelWomb!(*newPersonTemplate) |

        // Tell each offspring to reporduce (one fewer generation than the parent)
        cane!(generations - 1) |
        abel!(generations - 1) |

        // Make the individual offspring contracts
        // I want to do explicit substitution here
        *newPerson[cane / individual][caneWomb / individualWomb] |
        *newPerson[cane / individual][caneWomb / individualWomb]
      }
    }
  } |

  // Have adam reproduce for 3 generations
  // Creates 2^3 individuals
  adam!(3)
}
```

File: rholang/examples/intdupe.rho [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+new buildTree,
+    treeHelper,
+    stdout(`rho:io:stdout`) in {
+
+  /**
+   * Users can customize how many comms are desired by changing this line.
+   * Roughly 2^(n + 1) comms will occur. Use -1 for essentially unlimited comms.
+   * Other lines will not need to be changed by end users.
+   */
+  buildTree!(10) |
```
Comment (KentShikama) [inline rholang/examples/intdupe.rho]: Can we make the default 5 for now? We use random contracts from the examples folder for testnet right now and I don't want to heavily bias it towards stress testing contracts as of now. Of course in the future we probably want to pick out the contracts that we think would be more representative of what the real network would be running but I think that might be a couple weeks down the line.

Comment (lukasz-golebiewski) [reply rholang/examples/intdupe.rho]: I think this contract should go under `https://github.com/rchain/perf-harness/tree/master/contracts` @JoshOrndorff . If you opened a PR against that repo we could check right away what #COMMs is being generated

Comment (JoshOrndorff) [reply rholang/examples/intdupe.rho]: Sure, 5 sounds fine.

**Comment by JoshOrndorff** [comment]
Closing this in favor of posting it in the perf-harness repo. Now that I've seen that repo, I have some further questions about the scope of the work which I've posted in https://rchain.atlassian.net/browse/RCHAIN-2839

**Comment by JoshOrndorff** [comment]
Moved this work to https://github.com/rchain/perf-harness/pull/67

**Pull request closed without merging**